### PR TITLE
Generate a table of named vertices and edges from embeddings

### PR DESCRIPTION
**Pull request by bowu**
These changes aim to generate a table of named vertices in the pattern from embeddings. The table has vertex columns followed by edge columns. For example, if we have MATCH (a)-[x]->(b)-[y]->(c), the first three columns could be (a, b, c) depending on the match order, followed by (x, y).

### CODE REVIEW

File: libquery/src/MatchOp.cpp [hunk 1]
```diff
@@ -26,7 +26,30 @@ std::vector<Operator*> MatchOp::Compute() {
   QueryEmbeddingQueueType matched_embeddings = data_graph_->MatchQueryGraph(
       query_graph_, &query_vertex_to_matching_order);
 
-  TableBuilder builder         = TableBuilder::BuildFrom(matched_embeddings);
+  std::vector<std::pair<uint32_t, std::string>> node_name_pairs;
+  std::vector<std::pair<std::pair<uint32_t, uint32_t>, std::string>>
+      edge_name_pairs;
+
+  for (auto& p : var_name_to_node_index_) {
```
Comment (l-hoang) [inline libquery/src/MatchOp.cpp:33]: Add a comment saying this is pairing up embedding node index with a var name.

File: libquery/src/MatchOp.cpp [hunk 2]
```diff
@@ -26,7 +26,30 @@ std::vector<Operator*> MatchOp::Compute() {
...
+    unsigned embedding_node_id = query_vertex_to_matching_order[query_index];
+    node_name_pairs.emplace_back(embedding_node_id, name);
+  }
+
+  for (auto& p : edge_var_to_src_dst_pair_) {
+    std::string name = p.first;
+    uint32_t src_qid = p.second.first;
+    uint32_t dst_qid = p.second.second;
+    // matching order for those qids
+    uint32_t src_index = query_vertex_to_matching_order[src_qid];
```
Comment (l-hoang) [inline libquery/src/MatchOp.cpp:45]: Please prepend "embedding" or some indication in front of this variable to make it more clear that this is an index into the embedding.

File: libquery/src/MatchOp.cpp [hunk 3]
```diff
@@ -26,7 +26,30 @@ std::vector<Operator*> MatchOp::Compute() {
...
+      edge_name_pairs;
+
+  for (auto& p : var_name_to_node_index_) {
+    std::string name           = p.first;
+    uint32_t query_index       = p.second;
+    unsigned embedding_node_id = query_vertex_to_matching_order[query_index];
+    node_name_pairs.emplace_back(embedding_node_id, name);
+  }
+
+  for (auto& p : edge_var_to_src_dst_pair_) {
```
Comment (l-hoang) [inline libquery/src/MatchOp.cpp:40]: Add a comment about associating 2 embedding indices with an edge name.

File: libquery/src/Table.cpp [hunk 4]
```diff
@@ -84,27 +94,46 @@ void TableBuilder::EmplaceString(std::string value, size_t column_number) {
...
-  }
   TableSchema schema;
-  for (size_t col = 0; col < n_vertices; col++) {
-    std::ostringstream oss;
-    oss << "v" << col;
-    schema.AddColumn(oss.str(), arrow::Type::UINT32);
+  for (auto& p : node_name_pairs) {
+    schema.AddColumn(p.second, arrow::Type::UINT32);
+  }
+  // TODO: Do we always need uint64_t for edges?
```
Comment (l-hoang) [inline libquery/src/Table.cpp:110]: The current assumption is that edges that can fit on a single machine can indeed get to that size.

As a side note, this seems to be the only way to distinguish between a node variable and an edge variable at the moment, so I'd let it be unless you had another idea. (we need to be able to distinguish the 2 down in the op graph pipeline)

File: libquery/src/Table.cpp [hunk 5]
```diff
@@ -84,27 +94,46 @@ void TableBuilder::EmplaceString(std::string value, size_t column_number) {
...
+    for (auto& p : node_name_pairs) {
+      uint32_t v = emb.get_element(p.first).get_vid();
       builder.EmplaceUInt32(v, col);
       // builder.EmplaceInt64(v, col);
+      ++col;
+    }
+    for (auto& p : edge_name_pairs) {
+      uint32_t src    = p.first.first;
+      uint32_t dst    = p.first.second;
+      uint32_t src_id = emb.get_element(src).get_vid();
```
Comment (l-hoang) [inline libquery/src/Table.cpp:126]: Maybe prepend something to make it obvious is is a datagraph id?

File: libquery/src/Table.cpp [hunk 6]
```diff
@@ -84,27 +94,46 @@ void TableBuilder::EmplaceString(std::string value, size_t column_number) {
...
-    schema.AddColumn(oss.str(), arrow::Type::UINT32);
+  for (auto& p : node_name_pairs) {
+    schema.AddColumn(p.second, arrow::Type::UINT32);
+  }
+  // TODO: Do we always need uint64_t for edges?
+  for (auto& p : edge_name_pairs) {
+    schema.AddColumn(p.second, arrow::Type::UINT64);
   }
   TableBuilder builder(schema);
   for (auto& emb : embeddings) {
```
Comment (l-hoang) [inline libquery/src/Table.cpp:115]: A short comment about what this loop is doing would be helpful.

File: libcusp/include/galois/graphs/Partitioner.h [hunk 7]
```diff
@@ -1349,7 +1349,7 @@ class Partitioner {
           uint64_t cur =
               (lsrc > 0) ? dist_.large_array_out_indices_[lsrc - 1] : 0;
           for (auto edge : pview.edges(n)) {
-            auto gdst           = pview.GetEdgeDest(edge);
+            auto gdst                           = pview.GetEdgeDest(edge);
```
[the remark this answers is not in the document]
Comment (bowu) [reply libcusp/include/galois/graphs/Partitioner.h:1352]: Somehow git-clang-format touched it. I'll remove it.

File: libquery/include/querying/ops/ReturnOp.h [hunk 8]
```diff
@@ -11,7 +14,12 @@ class ReturnOp : public WithReturnBaseOp {
       : WithReturnBaseOp(Operator::OpType::kReturn, return_properties) {}
 
   // TODO(l-hoang) refine this
-  std::vector<Operator*> Compute() final { return PropagateOutput(); }
+  std::vector<Operator*> Compute() final {
+    for (auto& t : inputs_) {
+      std::cout << *dynamic_cast<Table*>(t.get());
+    }
+    return std::vector<Operator*>();
```
[the remark this answers is not in the document]
Comment (bowu) [reply libquery/include/querying/ops/ReturnOp.h:21]: Yes.

File: libquery/include/querying/Table.h [hunk 9]
```diff
@@ -29,10 +30,16 @@ class TableBuilder {
...
   void EmplaceUInt32(uint32_t value, size_t columnNumber);
+  void EmplaceUInt64(uint64_t value, size_t columnNumber);
   void EmplaceInt64(int64_t value, size_t columnNumber);
   void EmplaceDouble(double value, size_t columnNumber);
   void EmplaceString(std::string value, size_t columnNumber);
-  static TableBuilder BuildFrom(QueryEmbeddingQueueType&);
+  static TableBuilder
+  BuildFrom(QueryEmbeddingQueueType& embeddings,
+            std::vector<std::pair<uint32_t, std::string>>& node_name_pairs,
+            std::vector<std::pair<std::pair<uint32_t, uint32_t>, std::string>>&
```
Comment (danielmawhirter) [inline libquery/include/querying/Table.h:40]: may consider std::tuple instead of pair<pair>

**Comment by danielmawhirter** [comment]
Other than formatting, LGTM

**Pull request merged**